### PR TITLE
Div sqrt

### PR DESCRIPTION
**Pull request by midnighter95**
Add square root design.

Add DivSqrt Module in FP32 and its test.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -1,3 +1,9 @@
 [submodule "dependencies/chisel"]
 	path = dependencies/chisel
 	url = git@github.com:chipsalliance/chisel.git
+[submodule "berkeley-testfloat-3"]
+	path = berkeley-testfloat-3
+	url = git@github.com:ucb-bar/berkeley-testfloat-3.git
+[submodule "berkeley-softfloat-3"]
+	path = berkeley-softfloat-3
+	url = git@github.com:ucb-bar/berkeley-softfloat-3.git
```
Comment (sequencer) [inline .gitmodules:9]: @Avimitin can we use nix here?

File: tests/src/Tester.scala [hunk 2]
```diff
@@ -0,0 +1,192 @@
+package tests
```
Comment (sequencer) [inline tests/src/Tester.scala:1]: Move to `build.sc`, see T1/build.sc

File: tests/resources/includes/verilator.h [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+#define _ROCKET_VERILATOR_H
+
+#include "verilated_vcd_c.h"
+#include <stdlib.h>
+#include <stdio.h>
+
+extern bool verbose;
+extern bool done_reset;
+
+class VerilatedVcdFILE : public VerilatedVcdFile {
```
Comment (sequencer) [inline tests/resources/includes/verilator.h:11]: Use FST

File: tests/resources/csrc/vbridge_impl.h [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+struct testdata {
+  uint64_t a;
+  uint64_t b;
+  uint64_t expected_out;
+  function_t function;
+  exceptionFlag_t expectedException;
+};
+
+class VBridgeImpl {
```
Comment (sequencer) [inline tests/resources/csrc/vbridge_impl.h:40]: VBridge???

File: tests/resources/csrc/vbridge_impl.cc [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+        static_cast<exceptionFlag_t>(softfloat_exceptionFlags);
+    curData.expected_out = u.ui;
+
+    res.push_back(curData);
+  }
+
+  return res;
+}
+
+std::vector<testdata> mygen_az_f32(float32_t trueFunction(float32_t),
```
Comment (sequencer) [inline tests/resources/csrc/vbridge_impl.cc:145]: name...

File: tests/resources/csrc/vbridge_impl.cc [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+
+uint64_t VBridgeImpl::get_t() { return getCycle(); }
+
+int VBridgeImpl::timeoutCheck() {
+  if (terminate == true) {
+    LOG(INFO) << fmt::format("ran {} cases, cycles = {}", cnt - 1, get_t());
+    dpiFinish();
+  }
+  return 0;
+}
```
Comment (sequencer) [inline tests/resources/csrc/vbridge_impl.cc:18]: no need timeout check

File: common.sc [hunk 7]
```diff
@@ -37,3 +37,24 @@ trait ArithmeticModule
 }
 
 // TODO: migrate test to svsim
```
Comment (sequencer) [inline common.sc:39]: remove this todo

File: arithmetic/src/sqrt/QDS.scala [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    selectPoints,
+    a match {
+      case 2 =>
+        TruthTable(
+          Seq(
+            BitPat("b???0") -> BitPat("b10000"), //2
+            BitPat("b??01") -> BitPat("b01000"), //1
+            BitPat("b?011") -> BitPat("b00100"), //0
+            BitPat("b0111") -> BitPat("b00010") //-1
+          ),
```
Comment (sequencer) [inline arithmetic/src/sqrt/QDS.scala:68]: why defining this uop?

File: arithmetic/src/sqrt/QDS.scala [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+  * @param rWidth y Truncate width
+  * @param ohWidth quotient width
+  * @param partialDividerWidth equals to dTruncatedWidth - 1
+  */
+class QDS(rWidth: Int, ohWidth: Int, partialDividerWidth: Int, a: Int) extends Module {
+  // IO
+  val input = IO(Input(new QDSInput(rWidth, partialDividerWidth)))
+  val output = IO(Output(new QDSOutput(ohWidth)))
+
+  // from P269 in <Digital Arithmetic> : /16， should have got from SRTTable.
```
Comment (sequencer) [inline arithmetic/src/sqrt/QDS.scala:20]: TODO: calculate this table

File: arithmetic/src/float/RoundingUnit.scala [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+  val output = IO(Output(new Bundle{
+    val data = UInt(32.W)
+    val exceptionFlags = Output(Bits(5.W))
+  }))
+
+  val rmRNE = (input.roundingMode === RoundingMode.RNE)
+  val rmRTZ = (input.roundingMode === RoundingMode.RTZ)
+  val rmRDN = (input.roundingMode === RoundingMode.RDN)
+  val rmRUP = (input.roundingMode === RoundingMode.RUP)
+  val rmRMM = (input.roundingMode === RoundingMode.RMM)
```
Comment (sequencer) [inline arithmetic/src/float/RoundingUnit.scala:30]: decode out

**Review by sequencer — CHANGES_REQUESTED** [review]

**Review by cyyself — APPROVED** [review]
F32_DIV and SQRT with different rounding mode works. LGTM.

File: build.sc [hunk 11]
```diff
@@ -35,3 +48,221 @@ trait Arithmetic
...
+      "--timing",
+      "--main"
+    )
+  }
+
+  val testfloatBuildOutput = os.proc(Seq("nix", "build", ".#testfloat", "--no-link", "--print-out-paths"))
+    .call(os.pwd).out.text.trim()
+  val softfloatBuildOutput = os.proc(
+    Seq("nix", "build", ".#softfloat", "--no-link", "--print-out-paths"),
+  ).call(os.pwd).out.text.trim()
```
Comment (sequencer) [inline build.sc:161]: use env

**Review by sequencer — CHANGES_REQUESTED** [review]

**Review by sequencer — APPROVED** [review]

**Pull request merged**